File: apps/yukino/tests/tool-cache-marker.test.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

// Cache breakpoint placement.
//
// Expected behavior: the breakpoint lands on the last non-deferred tool. A tool carrying
// both defer_loading and cache_control causes the API to reject the entire request (400),
// and MCP tools are registered after built-in tools, so the array tail is often a deferred
// tool — we cannot simply mark the last element.
import { describe, it, expect } from "vitest";

import { markToolsForCache } from "@/llm/anthropic.js";
import type { ToolSchema } from "@/tools/types.js";

function marked(tools: ToolSchema[]): string[] {
  return tools.filter((t) => t.cache_control !== undefined).map((t) => t.name);
}
const rest: Omit<ToolSchema, "name"> = {
  description: "",
  input_schema: {
    type: "object",
    properties: {},
  },
};
describe("cache breakpoint placement", () => {
  it("scans backwards when the tail is a deferred tool", () => {
    const tools: ToolSchema[] = [
      { name: "ReadFile", ...rest },
      { name: "WriteFile", ...rest },
      { name: "ToolSearch", ...rest },
      { name: "mcp__linear__create_issue", defer_loading: true, ...rest },
      { name: "mcp__sentry__resolve", defer_loading: true, ...rest },
    ];
    markToolsForCache(tools);
    expect(marked(tools)).toEqual(["ToolSearch"]);
  });

  it("marks the last tool when none are deferred", () => {
    const tools: ToolSchema[] = [
      { name: "ReadFile", ...rest },
      { name: "Bash", ...rest },
    ];
    markToolsForCache(tools);
    expect(marked(tools)).toEqual(["Bash"]);
  });

  it("skips deferred tools interleaved in the middle", () => {
    const tools: ToolSchema[] = [
      { name: "Bash", ...rest },
      { name: "mcp__a__x", defer_loading: true, ...rest },
      { name: "Grep", ...rest },
      { name: "mcp__z__y", defer_loading: true, ...rest },
    ];
    markToolsForCache(tools);
    expect(marked(tools)).toEqual(["Grep"]);
  });

  it("marks nothing when all tools are deferred", () => {
    // The API requires at least one non-deferred tool; in practice built-in tools are
    // never deferred, so this is a defensive branch: skip caching rather than emit a
    // request that would be rejected with 400
    const tools: ToolSchema[] = [
      { name: "mcp__a__x", defer_loading: true, ...rest },
      { name: "mcp__b__y", defer_loading: true, ...rest },
    ];
    markToolsForCache(tools);
    expect(marked(tools)).toEqual([]);
  });

  it("does not throw on an empty array", () => {
    expect(() => {
      markToolsForCache([]);
    }).not.toThrow();
  });
});

File: yukino/llm/tool_cache_marker_test.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package llm

import (
	"testing"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/packages/param"
)

// Cache breakpoint placement.
//
// The marker must land on the last non-deferred tool. A tool carrying both defer_loading and
// cache_control causes the official endpoint to reject the entire request (400). MCP tools are
// registered after built-in tools, so after sorting the tail is often a deferred tool — simply
// marking the last element is not safe.
func TestMarkToolsForCache(t *testing.T) {
	plain := func(name string) anthropic.ToolUnionParam {
		return anthropic.ToolUnionParam{OfTool: &anthropic.ToolParam{Name: name}}
	}
	deferred := func(name string) anthropic.ToolUnionParam {
		return anthropic.ToolUnionParam{OfTool: &anthropic.ToolParam{
			Name:         name,
			DeferLoading: param.NewOpt(true),
		}}
	}
	// CacheControl is a value type; the zero value means unmarked. Comparing against the zero
	// value directly is the most reliable check.
	var unmarked anthropic.CacheControlEphemeralParam
	marked := func(tools []anthropic.ToolUnionParam) []string {
		var out []string
		for _, u := range tools {
			if u.OfTool == nil || u.OfTool.CacheControl == unmarked {
				continue
			}
			out = append(out, u.OfTool.Name)
		}
		return out
	}

	t.Run("scans backward when tail is deferred", func(t *testing.T) {
		tools := []anthropic.ToolUnionParam{
			plain("ReadFile"), plain("WriteFile"), plain("ToolSearch"),
			deferred("mcp__linear__create_issue"), deferred("mcp__sentry__resolve"),
		}
		markToolsForCache(tools)
		got := marked(tools)
		if len(got) != 1 || got[0] != "ToolSearch" {
			t.Fatalf("expected ToolSearch to be marked, got %v", got)
		}
	})

	t.Run("marks last tool when none are deferred", func(t *testing.T) {
		tools := []anthropic.ToolUnionParam{plain("ReadFile"), plain("Bash")}
		markToolsForCache(tools)
		if got := marked(tools); len(got) != 1 || got[0] != "Bash" {
			t.Fatalf("expected Bash to be marked, got %v", got)
		}
	})

	t.Run("skips deferred tools interleaved in the middle", func(t *testing.T) {
		tools := []anthropic.ToolUnionParam{
			plain("Bash"), deferred("mcp__a__x"), plain("Grep"), deferred("mcp__z__y"),
		}
		markToolsForCache(tools)
		if got := marked(tools); len(got) != 1 || got[0] != "Grep" {
			t.Fatalf("expected Grep to be marked, got %v", got)
		}
	})

	t.Run("marks nothing when all tools are deferred", func(t *testing.T) {
		// The official endpoint requires at least one non-deferred tool. In practice built-in
		// tools are never deferred, so this is a defensive branch: better to skip caching than
		// to send a request that will be rejected with 400.
		tools := []anthropic.ToolUnionParam{deferred("mcp__a__x"), deferred("mcp__b__y")}
		markToolsForCache(tools)
		if got := marked(tools); len(got) != 0 {
			t.Fatalf("expected no tools to be marked, got %v", got)
		}
	})
}
